### PR TITLE
JENKINS-58313 Add source commit and branch name

### PR DESCRIPTION
**Pull request by timja**
Background reading:
https://issues.jenkins-ci.org/browse/JENKINS-58313
https://github.com/jenkinsci/branch-api-plugin/pull/166
https://github.com/jenkinsci/branch-api-plugin/pull/163

Haven't figured out how to write a test for it yet, but will do so if required (pointers welcomed)

Basically trying to get two bit of information consistently:
- PR commit ID
- Branch name

Thoughts?

All names are up for debate

cc @bitwiseman @stephenc 

and FYI @henryju

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -97,7 +97,6 @@
             <groupId>org.jenkins-ci.plugins</groupId>
             <artifactId>branch-api</artifactId>
             <version>2.1.2</version>
-            <scope>test</scope>
```
Comment (stephenc) [inline pom.xml]: That you need to add a hard dependency on branch-api suggests this change should be an extension plugin and not in the core plugin. SCM implementations shouldn't be tied to orthogonal consumer concepts like branch-api

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java [hunk 2]
```diff
@@ -73,6 +74,9 @@ private static void createBuildCommitStatus(Run<?, ?> build, TaskListener listen
         SCMSource src = SCMSource.SourceByItem.findSource(build.getParent());
         SCMRevision revision = src != null ? SCMRevisionAction.getRevision(src, build) : null;
         if (revision != null) { // only notify if we have a revision to notify
+            GitHubEnvAction gitHubEnvAction = getGitHubEnvAction(revision);
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: Unnecessary to add this action.

https://github.com/jenkinsci/branch-api-plugin/blob/b875b2641d60e39f2b6074838f034a1b921269f6/src/main/java/jenkins/branch/BranchNameContributor.java#L55-L58

Followed by a call to https://github.com/jenkinsci/branch-api-plugin/blob/0e8545955a5ded2f482f77055ab064786267c710/src/main/java/jenkins/branch/BranchProjectFactory.java#L162 will get you the revision and then you can use the exact same logic for testing the instance types to extract the information you need.

A single class extension plugin with just an `EnvironmentContributor` implementation will do exactly what you are attempting without requiring changes to the core plugin

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java [hunk 3]
```diff
@@ -127,6 +131,31 @@ private static void createBuildCommitStatus(Run<?, ?> build, TaskListener listen
         }
     }
 
+    private static GitHubEnvAction getGitHubEnvAction(SCMRevision revision) {
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: I would suggest moving this logic to the environment contributor

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubEnvAction.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Map;
+import java.util.Objects;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Caches SCM values for a Run
+ */
+@Restricted(NoExternalUse.class)
+public class GitHubEnvAction extends InvisibleAction {
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubEnvAction.java]: Redundant class. Information is already retained the in the SCMRevision assoicated with the branch

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java (hunk 3 above)
Comment (timja) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: I don't have access to the revision there as far as I know.
It's possible to look it up, but that requires an API call from memory, and  this plugin is very focused on minimising API calls as it's easy to go over the hourly limit

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubEnvAction.java (hunk 4 above)
Comment (timja) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubEnvAction.java]: Yes but the javadoc for `EnvironmentContributor` says to use an `Action` to cache data that can be expensive to calculate, and it's expensive due to API calls required if you have to look it up

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java (hunk 2 above)
Comment (timja) [reply src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubBuildStatusNotification.java]: Will take a look thanks

**Comment by bitwiseman** [comment]
@timja I'm sorry, I'm kind of heads down on other work at the moment but I'd be interested in chatting about this and other changes you're working on this this area.  Do you Gitter or some other system?

**Comment by timja** [comment]
> @timja I'm sorry, I'm kind of heads down on other work at the moment but I'd be interested in chatting about this and other changes you're working on this this area. Do you Gitter or some other system?

Yes I'm timja on gitter

**Comment by timja** [comment]
> Recommend that this be a separate extension plugin. Nothing in the actually requires the changes be made in the github branch source plugin, rather a branch-api extension plugin that has a dependency on the github branch source (which, FTR should have been called the github-scm plugin) would be able to get the revision using the branch-api methods and contribute the environment variables

Thanks for the review, happy to make an extension plugin, will wait a bit to see if anyone else has any thoughts on whether this should live here or in another plugin

File: pom.xml [hunk 5]
```diff
@@ -98,7 +98,7 @@
...
+            <optional>true</optional>
```
Comment (stephenc) [inline pom.xml]: Optional is not a good plan. It complicates classloading. Should just be an extension plugin

**Comment by stephenc** [comment]
> I'm not seeing a particular reason to make this into a separate plugin.

SCM implementing plugins are not supposed to depend on branch-api (a SCM-api consumer). We typically have a test scope for interop testing, but a compile scope dependency (even if optional) significantly complexifies upgrades and migration.   Will not be anything near a fun time of this route is followed.

There is an alternative, namely augmenting SCM-API to provide a mechanism for SCMRevision to describe environment variables, then branch-api could consume that in its environment contributor and this PR would just report the env vars... absence that, this change is a major risk to future code evolution because it introduces a circular dependency, and there will be risk of `LinkageError`as branch-api evolves

**Pull request closed without merging**